compositor_pipeline/src/pipeline/output/whip.rs

```diff
@@ -0,0 +1,280 @@
+
+#[derive(Debug, Clone)]
+pub struct WhipCtx {
+    output_id: OutputId,
+    options: WhipSenderOptions,
+    sample_rate: u32,
+    stun_servers: Arc<Vec<String>>,
+    request_keyframe_sender: Option<Sender<()>>,
+    should_close: Arc<AtomicBool>,
+    tokio_rt: Arc<Runtime>,
+    event_emitter: Arc<EventEmitter>,
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum WhipError {
+    #[error("Bad status in WHIP response\nStatus: {0}\nBody: {1}")]
+    BadStatus(StatusCode, String),
+
+    #[error("WHIP request failed!\nMethod: {0}\nURL: {1}")]
+    RequestFailed(Method, Url),
+
+    #[error(
+        "Unable to get location endpoint, check correctness of WHIP endpoint and your Bearer token"
+    )]
+    MissingLocationHeader,
+
+    #[error("Invalid endpoint URL: {1}")]
+    InvalidEndpointUrl(#[source] ParseError, String),
+
+    #[error("Missing Host in endpoint URL")]
+    MissingHost,
+
+    #[error("Missing port in endpoint URL")]
+    MissingPort,
+
+    #[error("Failed to create RTC session description: {0}")]
+    RTCSessionDescriptionError(webrtc::Error),
+
+    #[error("Failed to set local description: {0}")]
+    LocalDescriptionError(webrtc::Error),
+
+    #[error("Failed to set remote description: {0}")]
+    RemoteDescriptionError(webrtc::Error),
+
+    #[error("Failed to parse {0} response body: {1}")]
+    BodyParsingError(&'static str, reqwest::Error),
+
+    #[error("Failed to create offer: {0}")]
+    OfferCreationError(webrtc::Error),
+
+    #[error(transparent)]
+    PeerConnectionInitError(#[from] webrtc::Error),
+
+    #[error("Failed to convert ICE candidate to JSON: {0}")]
+    IceCandidateToJsonError(webrtc::Error),
+
+    #[error(transparent)]
+    SerdeJsonError(#[from] serde_json::Error),
+
+    #[error(transparent)]
+    PayloadingError(#[from] PayloadingError),
+
+    #[error("Trickle ICE not supported")]
+    TrickleIceNotSupported,
+
+    #[error("Entity Tag missing")]
+    EntityTagMissing,
+
+    #[error("Entity Tag non-matching")]
+    EntityTagNonMatching,
+
+    #[error("Codec not supported: {0}")]
+    UnsupportedCodec(&'static str),
+}
+
+const WHIP_INIT_TIMEOUT: Duration = Duration::from_secs(60);
+
+impl WhipSender {
+    pub fn new(
+        output_id: &OutputId,
+        options: WhipSenderOptions,
+        packets_receiver: Receiver<EncoderOutputEvent>,
+        request_keyframe_sender: Option<Sender<()>>,
+        pipeline_ctx: &PipelineCtx,
+    ) -> Result<Self, OutputInitError> {
+        let payloader = Payloader::new(options.video, options.audio);
+        let packet_stream = PacketStream::new(packets_receiver, payloader, 1400);
+        let should_close = Arc::new(AtomicBool::new(false));
+        let (init_confirmation_sender, mut init_confirmation_receiver) =
+            oneshot::channel::<Result<(), WhipError>>();
+
+        let whip_ctx = WhipCtx {
+            output_id: output_id.clone(),
+            options: options.clone(),
+            sample_rate: pipeline_ctx.output_sample_rate,
```

Comment:
if you are passing that many fields from pipelient ctx maybe consider passing entire context instead